src/Streamly/Internal/Data/Array/Mut/Type.hs

```diff
@@ -0,0 +1,535 @@
+
+-- | Modify a given index of an array using a modifier function without checking
+-- the bounds.
+--
+-- Unsafe because it does not check the bounds of the array.
+--
+-- /Pre-release/
+modifyIndexUnsafe :: MonadIO m => Array a -> Int -> (a -> (a, b)) -> m b
+modifyIndexUnsafe Array {..} i f = do
+    liftInIO
+        $ \s# ->
+              case i + arrStart of
+                  I# n# ->
+                      case readArray# arrContents# n# s# of
+                          (# s1#, a #) ->
+                              let (a1, b) = f a
+                                  s2# = writeArray# arrContents# n# a1 s1#
+                               in (# s2#, b #)
+
+-- | Modify a given index of an array using a modifier function.
+--
+-- /Pre-release/
+modifyIndex :: MonadIO m => Array a -> Int -> (a -> (a, b)) -> m b
+modifyIndex arr@(Array {..}) i f = do
+    if i < arrLen && i > 0
+    then modifyIndexUnsafe arr i f
+    else invalidIndex "modifyIndex" i
+
+-------------------------------------------------------------------------------
+-- Resizing
+-------------------------------------------------------------------------------
+
+-- | Reallocates the array according to the new size. This is a safe function
+-- that always creates a new array and copies the old array into the new one. If
+-- the reallocated size is less then it results in a truncated version of the
+-- original array.
+--
+realloc :: MonadIO m => Int -> Array a -> m (Array a)
+realloc i arr = do
+    arr1 <- newArray i
+    let !newLen@(I# newLen#) = min i (arrLen arr)
+        !(I# arrS#) = arrStart arr
+        !(I# arr1S#) = arrStart arr1
+        arrC# = arrContents# arr
+        arr1C# = arrContents# arr1
+    liftInIO
+        $ \s# ->
+              let s1# = copyMutableArray# arrC# arrS# arr1C# arr1S# newLen# s#
+               in (# s1#, arr1 {arrLen = newLen, arrFree = i} #)
+
+reallocWith ::
+       MonadIO m => String -> (Int -> Int) -> Int -> Array a -> m (Array a)
+reallocWith label sizer reqSize arr = do
+    let oldSize = arrLen arr
+        newSize = sizer oldSize
+        safeSize = max newSize (oldSize + reqSize)
+    assert (newSize >= oldSize + reqSize || error badSize) (return ())
+    realloc safeSize arr
+
+    where
+
+    badSize = concat
+        [ label
+        , ": new array size is less than required size "
+        , show reqSize
+        , ". Please check the sizing function passed."
+        ]
+
+-------------------------------------------------------------------------------
+-- Snoc
+-------------------------------------------------------------------------------
+
+-- | Really really unsafe, appends the element into the first array, may
+-- cause silent data corruption or if you are lucky a segfault if the index
+-- is out of bounds.
+--
+-- /Internal/
+{-# INLINE snocUnsafe #-}
+snocUnsafe :: MonadIO m => Array a -> a -> m (Array a)
+snocUnsafe arr@Array {..} a = do
+    assert (arrLen + 1 <= arrFree) (return ())
```

Comment:
Assuming `arrFree` indicates the free index, this should be `arrLen <= arrFree` as the `arrLen` index is beyond the last index and it should be free.

Reply:
See comment above